src/index.ts

```diff
@@ -1,3 +1,149 @@
-export const cat = {
-  say: 'meow',
-};
+import { SCALE_FACTOR, TILEGRID_WORLD_CRS84 } from './constants';
+import { BoundingBox, LonLat, Tile, TileGrid } from './interfaces';
+import { Zoom } from './types';
+import { validateBoundingBox, validateLonlat, validateMetatile, validateTile, validateTileGrid, validateZoomLevel } from './validations';
+
+function clamp(value: number, minValue: number, maxValue: number): number {
+  if (value < minValue) {
+    return minValue;
+  }
+
+  if (value > maxValue) {
+    return maxValue;
+  }
+
+  return value;
+}
+
+function tileHeight(zoom: Zoom, referenceTileGrid: TileGrid): number {
+  return (
+    (referenceTileGrid.boundingBox.north - referenceTileGrid.boundingBox.south) / (referenceTileGrid.numberOfMinLevelTilesY * SCALE_FACTOR ** zoom)
+  );
+}
+
+function tileWidth(zoom: Zoom, referenceTileGrid: TileGrid): number {
+  return (
+    (referenceTileGrid.boundingBox.east - referenceTileGrid.boundingBox.west) / (referenceTileGrid.numberOfMinLevelTilesX * SCALE_FACTOR ** zoom)
+  );
+}
+
+function* tilesGenerator(limits: [number, number, number, number], zoom: Zoom, metatile: number): Generator<Tile, undefined, undefined> {
+  for (let y = limits[1]; y <= limits[3]; y++) {
+    for (let x = limits[0]; x <= limits[2]; x++) {
+      yield { x: x, y: y, z: zoom, metatile: metatile };
```

Comment:
1. what is the reasoning behind returning the metatile if its unchanged and unused?
2. using the shortened method for property value will make it more readable https://attacomsian.com/blog/javascript-object-property-shorthand

Reply:
1. it should be part of the tile because it affects the x and y numbering. for example in the case you received a tile you need to know which metatile number was involved in it's creation
2 . ok